### PR TITLE
initial example created

### PR DESCRIPTION
**Pull request by galmail**

### CODE REVIEW

File: next-pino/logger.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+const pinoConfig = {
+  browser: {
+    asObject: true
+  }
+}
+
+if (config.serverUrl) {
+  pinoConfig.browser.transmit = {
+    level: 'info',
+    send: async (level, logEvent) => {
```
Comment (mcollina) [inline next-pino/logger.js:18]:
```suggestion
    send (level, logEvent) {
```

I don't think this should be async.

File: pages/api/hello.js [hunk 2]
```diff
@@ -0,0 +1,7 @@
+const logger = require('pino')()
```
Comment (mcollina) [inline pages/api/hello.js:1]: This is not needed, you should see `req.log`  as a logger in the API route.

File: package.json [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+  "name": "my-app",
+  "version": "0.1.0",
+  "private": true,
+  "scripts": {
+    "dev": "node server.js",
+    "build": "next build",
+    "start": "NODE_ENV=production node server.js"
+  },
+  "dependencies": {
+    "next": "9.4.0",
```
Comment (mcollina) [inline package.json:11]:
```suggestion
    "next": "^9.4.0",
```

File: pages/index.js [hunk 4]
```diff
@@ -0,0 +1,195 @@
+import Head from 'next/head'
+import logger from '../next-pino/logger'
```
Comment (mcollina) [inline pages/index.js:2]: This logger should be the one attached to the request when doing SSR.
Have we got access to `req` and `res` inside the component?

Comment (galmail) [reply pages/index.js:2]: The thing is that once the component is rendered and is served to the browser, we want to be able to send the logs to the server and for that we will need:

`navigator.sendBeacon(${config.serverUrl}/log, blob)`

Comment (mcollina) [reply pages/index.js:2]: I know. We need a bit of conditionals in here: if there is `req` somewhere, use that. If we are in the browser, use the one with sendBeacon.

File: next-pino/logger.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
+import pino from 'pino'
+
```
Comment (mcollina) [inline next-pino/logger.js:2]: Please add a comment explaining this is the logger for the browser.

File: server.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+    if (pathname === '/log') {
+        let data = null
+        req.on('data', chunk => {
+            data = chunk
+        })
+        req.on('end', () => {
+            const { msg, level = 'info' } = JSON.parse(data)
+            req.log[level](msg)
+        })
```
Comment (mcollina) [inline server.js:29]: Please use the `stream.finished()` utility inside end.

File: server.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        res.statusCode = 200
+        res.setHeader('Content-Type', 'text/plain')
+        res.end('logged on server')
```
Comment (mcollina) [inline server.js:32]: please send the response only after we have recevied the body.

Comment (galmail) [reply server.js]: ok I've made the change, have a look and let me know.

File: server.js (hunk 6 above)
Comment (galmail) [reply server.js:38]: I've added the ok response inside that block, but I'm not sure how to use `stream.finished()` ?

Comment (mcollina) [reply server.js:38]: https://nodejs.org/api/stream.html#stream_stream_finished_stream_options_callback

File: server.js [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    // Be sure to pass `true` as the second argument to `url.parse`.
+    // This tells it to parse the query portion of the URL.
+    const parsedUrl = parse(req.url, true)
+    const { pathname, query } = parsedUrl
+    logger(req, res)
+
+    if (pathname === '/log') {
+        let data = null
+        req.on('data', chunk => {
+            data = chunk
```
Comment (mcollina) [inline server.js:24]:
```suggestion
            data += chunk
```

File: server.js [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+app.prepare().then(() => {
+  createServer((req, res) => {
+    // Be sure to pass `true` as the second argument to `url.parse`.
+    // This tells it to parse the query portion of the URL.
+    const parsedUrl = parse(req.url, true)
+    const { pathname, query } = parsedUrl
+    logger(req, res)
+
+    if (pathname === '/log') {
+        let data = null
```
Comment (mcollina) [inline server.js:22]: You need to do `req.setEncoding('utf8')`.

File: server.js [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+            data = chunk
+        })
+        req.on('error', err => {
+            req.log.error(err)
+            res.statusCode = 500
+            res.setHeader('Content-Type', 'text/plain')
+            res.end('error ocurred when logging to server')
+        })
+        req.on('end', () => {
+            const { msg, level = 'info' } = JSON.parse(data)
```
Comment (mcollina) [inline server.js:33]: Please ad d a try catch around this.

File: server.js (hunk 6 above)
Comment (galmail) [reply server.js:53]: done, thanks!

**Comment by galmail** [comment]
@mcollina I've made the changes, have a look and let me know, thanks.

**Pull request merged**